File: app/src/main/java/com/lucas/omnia/activities/NewCommentActivity.java
package com.lucas.omnia.activities;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.ServerValue;
import com.google.firebase.database.Transaction;
import com.google.firebase.database.ValueEventListener;
import com.lucas.omnia.R;
import com.lucas.omnia.databinding.ActivityNewCommentBinding;
import com.lucas.omnia.models.Comment;
import com.lucas.omnia.models.Post;
import com.lucas.omnia.models.User;

import java.util.HashMap;
import java.util.Map;

import static com.lucas.omnia.activities.CommentsActivity.commentsReference;
import static com.lucas.omnia.activities.CommentsActivity.postKey;
import static com.lucas.omnia.activities.CommentsActivity.postReference;

public class NewCommentActivity extends BaseActivity {

    private static final String TAG = "NewCommentActivity";
    private static final String REQUIRED = "Required";

    private DatabaseReference databaseReference;

    private ActivityNewCommentBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityNewCommentBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        databaseReference = getDatabaseReference();

        binding.newCommentFabSubmit.setOnClickListener(v -> submitComment());
    }

    private void submitComment() {
        final String body = binding.newCommentEtBody.getText().toString();

        // Body is required
        if (TextUtils.isEmpty(body)) {
            binding.newCommentEtBody.setError(REQUIRED);
            return;
        }

        // Disable button so there are no multi-posts
        setEditingEnabled(false);
        Toast.makeText(this, getString(R.string.new_comment_toast_commenting), Toast.LENGTH_SHORT).show();

        final String userId = getUid();
        databaseReference.child("users").child(userId).addListenerForSingleValueEvent(
                new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        User user = dataSnapshot.getValue(User.class);

                        if (user == null) {
                            Log.e(TAG, "User " + userId + " is unexpectedly null");
                            Toast.makeText(NewCommentActivity.this,
                                    getString(R.string.new_post_toast_user_fetch_error),
                                    Toast.LENGTH_SHORT).show();
                        } else {
                            writeNewComment(userId, user.getUsername(), body);
                            incrementCommentsCount();
                        }
                        setEditingEnabled(true);
                        finish();
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {
                        Log.e(TAG, "getUser:onCancelled", databaseError.toException());
                        setEditingEnabled(true);
                    }
                });
    }

    private void setEditingEnabled(boolean enabled) {
        binding.newCommentEtBody.setEnabled(enabled);
        if (enabled) {
            binding.newCommentFabSubmit.show();
        } else {
            binding.newCommentFabSubmit.hide();
        }
    }

    private void writeNewComment(String userId, String username, String body) {
        // Create new item_comment at /post-comments
        String key = commentsReference.push().getKey();
        Comment comment = new Comment(userId, username, body);
        Map<String, Object> commentValues = comment.toMap();

        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("/post-comments/" + postKey + "/" + key, commentValues);
        databaseReference.updateChildren(childUpdates);

        // Set timestamp
        databaseReference.child("post-comments").child(postKey).child(key).child("timestamp").setValue(ServerValue.TIMESTAMP);
    }

    private void incrementCommentsCount() {
        postReference.runTransaction(new Transaction.Handler() {
            @NonNull
            @Override
            public Transaction.Result doTransaction(@NonNull MutableData mutableData) {
                Post p = mutableData.getValue(Post.class);
                if (p == null) {
                    return Transaction.success(mutableData);
                }

                // Increment commentsCount of item_post
                p.commentCount = p.commentCount + 1;

                // Set value and report transaction success
                mutableData.setValue(p);
                return Transaction.success(mutableData);
            }

            @Override
            public void onComplete(DatabaseError databaseError, boolean committed,
                                   DataSnapshot currentData) {
                // Transaction completed
                Log.d(TAG, "CommentTransaction:onComplete:" + databaseError);
            }
        });
    }
}

File: app/src/main/java/com/lucas/omnia/activities/NewCommentActivity.kt
package com.lucas.omnia.activities

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.widget.Toast
import com.google.firebase.database.*
import com.lucas.omnia.R
import com.lucas.omnia.databinding.ActivityNewCommentBinding
import com.lucas.omnia.models.Comment
import com.lucas.omnia.models.Post
import com.lucas.omnia.models.User
import java.util.*

class NewCommentActivity : BaseActivity() {
    private var databaseRef: DatabaseReference? = null
    private var binding: ActivityNewCommentBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityNewCommentBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        databaseRef = databaseReference
        binding!!.newCommentFabSubmit.setOnClickListener { submitComment() }
    }

    private fun submitComment() {
        val body = binding!!.newCommentEtBody.text.toString()

        // Body is required
        if (TextUtils.isEmpty(body)) {
            binding!!.newCommentEtBody.error = REQUIRED
            return
        }

        // Disable button so there are no multi-posts
        setEditingEnabled(false)
        Toast.makeText(this, getString(R.string.new_comment_toast_commenting), Toast.LENGTH_SHORT).show()
        val userId = uid
        databaseRef!!.child("users").child(userId).addListenerForSingleValueEvent(
                object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val user = dataSnapshot.getValue(User::class.java)
                        if (user == null) {
                            Log.e(TAG, "User $userId is unexpectedly null")
                            Toast.makeText(this@NewCommentActivity,
                                    getString(R.string.new_post_toast_user_fetch_error),
                                    Toast.LENGTH_SHORT).show()
                        } else {
                            writeNewComment(userId, user.username, body)
                            incrementCommentsCount()
                        }
                        setEditingEnabled(true)
                        finish()
                    }

                    override fun onCancelled(databaseError: DatabaseError) {
                        Log.e(TAG, "getUser:onCancelled", databaseError.toException())
                        setEditingEnabled(true)
                    }
                })
    }

    private fun setEditingEnabled(enabled: Boolean) {
        binding!!.newCommentEtBody.isEnabled = enabled
        if (enabled) {
            binding!!.newCommentFabSubmit.show()
        } else {
            binding!!.newCommentFabSubmit.hide()
        }
    }

    private fun writeNewComment(userId: String, username: String?, body: String) {
        // Create new item_comment at /post-comments
        val key = CommentsActivity.commentsReference!!.push().key
        val comment = Comment(userId, username, body)
        val commentValues = comment.toMap()
        val childUpdates: MutableMap<String, Any> = HashMap()
        childUpdates["/post-comments/" + CommentsActivity.postKey + "/" + key] = commentValues
        databaseRef!!.updateChildren(childUpdates)

        // Set timestamp
        databaseRef!!.child("post-comments").child(CommentsActivity.postKey!!).child(key!!).child("timestamp").setValue(ServerValue.TIMESTAMP)
    }

    private fun incrementCommentsCount() {
        CommentsActivity.postReference!!.runTransaction(object : Transaction.Handler {
            override fun doTransaction(mutableData: MutableData): Transaction.Result {
                val p = mutableData.getValue(Post::class.java)
                        ?: return Transaction.success(mutableData)

                // Increment commentsCount of item_post
                p.commentCount = p.commentCount + 1

                // Set value and report transaction success
                mutableData.value = p
                return Transaction.success(mutableData)
            }

            override fun onComplete(databaseError: DatabaseError?, committed: Boolean,
                                    currentData: DataSnapshot?) {
                // Transaction completed
                Log.d(TAG, "CommentTransaction:onComplete:$databaseError")
            }
        })
    }

    companion object {
        private const val TAG = "NewCommentActivity"
        private const val REQUIRED = "Required"
    }
}